edgehog-device-runtime-forwarder/src/connections.rs

```diff
@@ -0,0 +1,362 @@
+
+//! Handle the interaction between the device connections and the bridge.
+
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+use std::ops::ControlFlow;
+use std::time::Duration;
+
+use backoff::ExponentialBackoff;
+use displaydoc::Display;
+use futures::{future, SinkExt, StreamExt, TryFutureExt};
+use thiserror::Error as ThisError;
+use tokio::net::TcpStream;
+use tokio::select;
+use tokio::sync::mpsc::{unbounded_channel, UnboundedReceiver, UnboundedSender};
+use tokio::time::timeout;
+use tokio_tungstenite::{
+    connect_async, tungstenite::Error as TungError, tungstenite::Message as TungMessage,
+    MaybeTlsStream, WebSocketStream,
+};
+
+use tracing::{debug, error, info, instrument, trace, warn};
+use url::Url;
+
+use crate::connection::{Connection, ConnectionError, ConnectionHandle};
+use crate::messages::{Http, HttpRequest, Id, ProtoError, ProtoMessage, Protocol as ProtoProtocol};
+
+/// WebSocket stream alias.
+pub type WsStream = WebSocketStream<MaybeTlsStream<TcpStream>>;
+
+/// The time interval in seconds that the device will wait before sending a Ping frame if no data is received from the WebSocket.
+pub const PING_TIMEOUT: u64 = 5;
+
+/// Errors occurring during the connections management.
+#[derive(Display, ThisError, Debug)]
+#[non_exhaustive]
+pub enum Error {
+    /// Error performing exponential backoff when trying to (re)connect with the bridge.
+    WebSocket(#[from] TungError),
+    /// Protobuf error.
+    Protobuf(#[from] ProtoError),
+    /// Connection error.
+    Connection(#[from] ConnectionError),
+    /// Wrong message.
+    WrongMessage,
+    /// The connection does not exists, id: `{0:?}`.
+    ConnectionNotFound(Id),
+    /// Connection ID already in use, id: `{0:?}`.
+    IdAlreadyUsed(Id),
+}
+
+/// Handler responsible for establishing a websocket connection between a device and the bridge
+/// and for receiving and sending data from/to it.
+#[derive(Debug)]
+pub struct ConnectionsManager {
+    /// Collection of connections, each identified by an ID.
+    connections: Connections,
+    /// Websocket stream between the device and the bridge.
+    ws_stream: WsStream,
+    /// Channel used to send through the websocket messages coming from each connection.
+    rx_ws: UnboundedReceiver<ProtoMessage>,
+    /// bridge URL.
+    url: Url,
+}
+
+impl ConnectionsManager {
+    /// Establish a new WebSocket connection between the device and the bridge.
+    #[instrument]
+    pub async fn connect(url: Url) -> Result<Self, Error> {
+        // TODO: check if, when a wrong URL is passed, it will endlessly try to connect
+
+        let ws_stream = Self::ws_connect(&url).await?;
+
+        // this channel is used by tasks associated to each connection to communicate new
+        // information available on a given websocket between the device and TTYD.
+        // it is also used to forward the incoming data from TTYD to the device.
+        let (tx_ws, rx_ws) = unbounded_channel();
```

Comment:
Set a bound to the maximum number of connection's messages

Reply:
What do you think a good number of messages inside the channel could be?

Reply:
In the astarte_device_sdk we use 50 as an arbitrary value

Reply:
Another strategy could be using the crate `num_cpu`;

Reply:
Use just `num_cpu` could be provide a strange behavior on board with single or double CPU, maybe we could add a multiplier by `num_cpu`, for instance:
`12 * num_cpu`